### DISCUSSION TITLE
What is the recommended approach for migrating from local storage to S3?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#4048] [Q&A]
Suppose I have a PocketBase installation storing `/api/files` to local storage. Now, I want to migrate it to S3.

How would I do that, and what will happen if I tell PocketBase to begin using S3 for asset storage and retrieval after I already have assets in local storage?

**Comment by ganigeorgiev** [comment]
You'll have to migrate your local `pb_data/storage` files to the S3 bucket. This can be done using tools like [rclone](https://github.com/rclone/rclone), [s5cmd](https://github.com/peak/s5cmd) or other similar cli tools (note that the directory structure need to be preserved as object prefix in this case, aka. the full S3 filename would be something like ":collectionId/:recordId/:filename").

In both cases when `/api/files` is used the files are retrieved and served via the PocketBase server.

**Reply by benallfree** [reply]
Hmm... how feasible would it be to write a migration tool into the PB admin for this? I can see a lot of PocketHost users running into this problem where they start small with local storage and then realize they need cloud storage. 

I could write it in the PH admin, but correctly transferring and naming everything seems tightly coupled to PB's internal scheme.

**Reply by ganigeorgiev** [reply]
You don't have to name anything if you use the above tools.
`rclone` at least will preserve by default your local directory structure and will "transform" the directories path into a S3 blob filename prefix.

There are no plans for now to integrate this in PocketBase as there are too many edge cases, it is not something that users have to do frequently and there are already better existing tools that helps with this and other storage migration tasks.
